text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+some catch all modules like `sentry.models` which import everything.  This means that it's hard for as to track
+the actual dependencies from different pieces of code and also hides away circular imports.  One of the results of
+this is that if you run a test against a single pytest module, you still pull in the entire codebase which already
+takes a couple of seconds.  It also means that it's harder for test infrastructure code to analyze the minimal set
+of dependencies that might need testing.
+
+Likewise code in `sentry.utils` should most likely no longer import models etc.  Some of that code might event
+move out of `sentry` entirely.
+
+### Move Code from Getsentry to Senty
+
+Currently quite a bit of code lives in `getsentry` for reasons of assumed convenience.  However this has created
+the situation that it is quite easy to ship regressions because the changes in `getsentry` were not considered.
+There is likely a whole range of code that does not need to live in `getsentry` and moving it to `sentry` would
+make the testsuite more reliable, faster to run and catch more regressions early.
+
+### Remove Pickle in Database
+
+We still use pickled models in database code quite extensively.  This reliance on pickle makes it harder than
+necessary to read and write from this outside of the main Python codebase.  Moving this to JSON will simplify the
+handling of moving such code into environments where other code wants to access it.  We already require depickling
+in the data pipeline for such cases.
+
+### Phase out Buffers where Possible
+
+Buffers as they exist today could in many cases be replaced by better uses of Kafka and clickhouse.  Some buffers are
+not even as important any more for the product.  For instance the total counts for issues are already hidden from the
+more prominent places in the UI.  The issue with buffers is that they are rather hard to scale, require the use of
+pickle for the model updates and are hard to work with when it comes to filtering.  For instance we are able to give
+you the total event count on an issue, but not when broken down by some of the filters that the UI wants to provide
+(environment and release is possible, but any tag search will not able to give the right counts).
+
+## Relay
+
+Within relay some changes are likely to be necessary for continued to support of providing a stable service.  In
+particular as more and more work moves into Relay certain risks that the current architecture is carrying need to
+be addressed.  The goals here are horizontal scalability, the ability to route data more efficiently and to recover
+from catastrophic failures.
+
+### Traffic Steering
+
+As we are now performing aggregations in Relay it's beneficial to be able to route traffic intelligently through the
+layers.  For instance we can achieve much better aggregations by ensuring that data from related metrics keys are
+forwarded to the same processing relays.
```

Comment:
I think there are three types of traffic routing problems we are facing today, though they are really instances of the same problem and we could have one system to solve the three. 

1) Traffic affinity. Which is what you mentioned.
2) Partitioning by organization to run parallel slices of the architecture. Everything based on Kafka does not scale indefinitely horizontally on one topic. We generally first scale by adding partitions, then by adding parallelism on the consumer, but at a point (above 64 partitions) break the traffic into independent topics and partition organizations across topics. As we scale more this last scenario will happen more.
3) Smart semantic partitioning. Besides Relay metrics aggregations there is a number of Kafka consumers that become more efficient if traffic is sort of semantic partitioned. As an example: if, tentatively, an organization will be only present in a small number of partitions. The gain is still cache locality.

I think the problem is always the same:
- map an organization to a logical shard.
- have a component (that can even be a service) that dynamic ally or statically route a logical shard to a physical resource (a relay node, a kafka topic or a kafka partition, etc.)
- have a feedback loop for dynamic traffic routing that observes the load on the physical resource and inform the component that performs the routing.

I think there is some value in building this as a Relay agnostic system that works based on an abstract concept of partition and physical resource.

I wrote down some of that in the comments here 
https://www.notion.so/sentry/Traffic-Steering-f2244ea8c40d405ebca998fe8dc6fd05
This is the scenario in the case 2 above https://www.notion.so/sentry/Traffic-routing-between-shards-d5f15d5e419f4dc5aac67c1c1140fa98

Reply:
I agree with filippo and I wish we would start building this service sooner rather than later. Our current initiatives don't have a common interface to code against, and since multiple teams are working on sharding/slicing/hybrid-cloud solutions I fear that there will be some avoidable fragmentation in approaches.